effectful/handlers/torch.py

```diff
@@ -180,20 +170,70 @@ def reindex_flat_tensor(t):
             return t
 
         result = t.reshape(inds[0].shape + t.shape[1:])
-        return torch_getitem(result, tuple(var() for (var, _) in reindex_fvs))
+        return torch_getitem(result, tuple(k() for k in sized_fvs.keys()))
+
+    result = tree.map_structure(reindex_flat_tensor, flat_result)
+    return result
+
+
+@defop
+def to_tensor_(
+    t: Annotated[torch.Tensor, Scoped[A | B]],
+    *args: Annotated[Operation[[], torch.Tensor], Scoped[A]],
+) -> Annotated[torch.Tensor, Scoped[B]]:
+    def _evaluate(expr):
+        if isinstance(expr, Term):
+            (args, kwargs) = tree.map_structure(_evaluate, (expr.args, expr.kwargs))
+            return _partial_eval(expr)
+        if tree.is_nested(expr):
+            return tree.map_structure(_evaluate, expr)
+        return expr
+
+    if not isinstance(t, Term):
+        return t
 
-    return tree.map_structure(reindex_flat_tensor, flat_result)
+    result = _evaluate(t)
+    if not isinstance(result, Term) or not args:
+        return result
+
+    # ensure that the result is a torch_getitem with a tensor as the first argument
+    if not (result.op is torch_getitem and isinstance(result.args[0], torch.Tensor)):
+        raise NotImplementedError
+
+    tensor = result.args[0]
+    dims = result.args[1]
+    assert isinstance(dims, Sequence)
+
+    # ensure that the order is a subset of the named dimensions
+    order_set = set(args)
+    if not order_set <= set(a.op for a in dims if isinstance(a, Term)):
+        raise NotImplementedError
+
+    # permute the inner tensor so that the leading dimensions are in the order
+    # specified and the trailing dimensions are the remaining named dimensions
+    # (or slices)
+    reindex_dims = [
+        i
+        for i, o in enumerate(dims)
+        if not isinstance(o, Term) or o.op not in order_set
+    ]
+    dim_ops = [a.op if isinstance(a, Term) else None for a in dims]
+    perm = [dim_ops.index(o) for o in args] + reindex_dims
+    tensor = tensor.permute(perm)
+    return tensor[(slice(None),) * len(args) + tuple(dims[i] for i in reindex_dims)]
 
 
-def to_tensor(t: T, order: Collection[Operation[[], torch.Tensor]] | None = None) -> T:
+def to_tensor(
```

Comment:
Why do we need both `to_tensor` and `to_tensor_` instead of just a single `to_tensor` operation?

Reply:
I didn't think that `Scoped` annotations worked inside collection types, so `to_tensor_` needed to have a different interface than `to_tensor`. I think merging them would be fine, but would change the API slightly.

Reply:
OK, I'm not seeing why we'd need the `to_tensor` interface at all in that case. The variadic `to_tensor_` signature seems nicer anyway since it's closer to `deffn`'s.

Reply:
Let's use the variadic interface then.